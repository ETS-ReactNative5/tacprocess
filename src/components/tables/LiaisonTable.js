import React from 'react';
import propTypes from "prop-types";
import '../../styles/components/tables.css';
import {getSaltAstronomerName, getSaltAstronomerUsername} from '../../util/salt-astronomer';
import {compareByProposalCode} from '../../util/proposal'
import {isLiaisonAstronomerUpdated} from '../../util/proposal-filtering'

const getEventValue = (event, astronomers) => {
  event.preventDefault()
  if (event.target.name === 'select') return getSaltAstronomerUsername(event.target.value, astronomers)
  return event.target.value
}

const LiaisonTable = ({proposals, canAssign, selectArray, requestSummary, username, setLiaison, initProposals, semester}) => (
  <div className='SATableDiv'>
    <h1>Salt Astronomers Liaison Assigning</h1>
    <table className='SATable' align='center'>
      <thead>
      <tr>
        <th>Proposal Code</th>
        <th>Summary</th>
        <th>Proposal Title</th>
        <th>Proposal Investigator</th>
        <th>Liaison Astronomer</th>
      </tr>
      </thead>
      <tbody>
      {
        proposals.sort(compareByProposalCode).map(p => {
          const liaison = getSaltAstronomerName(p.liaisonAstronomer, selectArray)
          const col = isLiaisonAstronomerUpdated(p, initProposals) ? {color: 'blue'} : {color: 'black'}
          return ( <tr key={`liaison-${p.proposalCode}`}>
          <td>
            <a target="_blank"
               href={`https://www.salt.ac.za/wm/proposal/${p.proposalCode}`}>
              {p.proposalCode}
            </a>
          </td>
          <td className="width-100">
            <a className="file-download"
               href=""
               onClick={e => requestSummary(e, p.proposalCode, semester)}>
              Download
            </a>
          </td>
          <td className=" table-height width-400">{p.title}</td>
          <td className="width-100">{p.pi}</td>
          {
            canAssign ?
              <td>
                <select style={col} defaultValue={liaison} onChange={
                  e => setLiaison(getEventValue(e, selectArray), p.proposalCode, true)
                } name={'selector'}>
                  {
                    !liaison && <option>none</option>
                  }
                  {
                    selectArray.map(op => (
                      <option key={op.username} value={op.name} name={op.username}>
                        {op.name}
                      </option>
                    ))
                  }

                </select>
              </td> :
              <td>
                {
                  <div>
                    <input
                      checked={p.liaisonAstronomer}
                      disabled={initProposals.filter(
                        ip => ip.proposalCode === p.proposalCode)[0].liaisonAstronomer
                      }
                      type={"checkbox"}
                      value={ username }
                      onChange={e => setLiaison(getEventValue(e, selectArray), p.proposalCode, e.target.checked)}/>
                    { p.liaisonAstronomer ?
                      <a style={col}> {liaison} </a>: <a style={{color: 'red'}}>{'Select'}</a>}
                  </div>
                }
              </td>
          }

        </tr>)})
      }

      </tbody>
    </table>
  </div>
  )

LiaisonTable.propTypes = {
  proposals: propTypes.array.isRequired,
  initProposals: propTypes.array.isRequired,
  canAssign: propTypes.bool.isRequired,
  selectArray: propTypes.array.isRequired,
  username: propTypes.string.isRequired,
  requestSummary: propTypes.func.isRequired,  // todo request summary should know the current selected semester
  setLiaison: propTypes.func.isRequired,
  semester: propTypes.string.isRequired
};
export default LiaisonTable
