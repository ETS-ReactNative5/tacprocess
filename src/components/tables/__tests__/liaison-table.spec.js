import React from "react"
import renderer from "react-test-renderer"
import { shallow, mount, render } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import LiaisonTable from "../LiaisonTable"

const proposals = [
  {
    proposalId: '0',
    proposalCode: '2018-1-SCI-038',
    title: 'Surface Chemistry of Peculiar Hot Subdwarfs',
    pi: 'Jeffery Simon',
    liaisonAstronomer: 'Eric'
  },
  {
    proposalId: '1',
    proposalCode: '2015-2-MLT-006',
    title: 'THE SALT GRAVITATIONAL LENSING LEGACY SURVEY',
    pi: 'Serjeant Stephen',
    liaisonAstronomer: 'brent'
  }
]

const canAssign = [true, false]

const selectArray = [
  {
    username: 'brent',
    name: 'brent'
  },
  {
    username: 'encarni',
    name: 'encarni'
  }
]

const requestSummary = () => {

}

// Checking if the LiaisonTable Component renders correctly for different input
describe("LiaisonTable Component", () => {

  //Enzyme method testing if it renders correctly with empty values of the proposals
  it("Render the LiaisonTable Component having unpopulated props with no errors", () => {
    const rendered = mount(<LiaisonTable proposals= {[]} canAssign= {canAssign[1]} selectArray= {[]} requestSummary= {requestSummary} />)
    expect(shallowToJson(rendered)).toMatchSnapshot()
  })

  it("Render the LiaisonTable Component having populated props with no errors", () => {
    const rendered = mount(<LiaisonTable proposals= {proposals} canAssign= {canAssign[0]} selectArray= {selectArray} requestSummary= {requestSummary} />)
    expect(shallowToJson(rendered)).toMatchSnapshot()
  })
})
