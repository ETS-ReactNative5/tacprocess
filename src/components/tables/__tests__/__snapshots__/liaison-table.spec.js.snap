// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiaisonTable Component Render the LiaisonTable Component having populated props with no errors 1`] = `
<LiaisonTable
  canAssign={true}
  initProposals={
    Array [
      Object {
        "liaisonAstronomer": "Proposal Liaison Astronomer",
        "pi": "Proposal Investigator",
        "proposalCode": "Proposal Code",
        "proposalId": "Proposal Id",
        "title": "Proposal Title",
      },
    ]
  }
  proposals={
    Array [
      Object {
        "liaisonAstronomer": "brent",
        "pi": "Serjeant Stephen",
        "proposalCode": "2015-2-MLT-006",
        "proposalId": "1",
        "title": "THE SALT GRAVITATIONAL LENSING LEGACY SURVEY",
      },
      Object {
        "liaisonAstronomer": "Eric",
        "pi": "Jeffery Simon",
        "proposalCode": "2018-1-SCI-038",
        "proposalId": "0",
        "title": "Surface Chemistry of Peculiar Hot Subdwarfs",
      },
    ]
  }
  requestSummary={[Function]}
  selectArray={
    Array [
      Object {
        "name": "brent",
        "username": "brent",
      },
      Object {
        "name": "encarni",
        "username": "encarni",
      },
    ]
  }
  semester="2018-1"
  setLiaison={[Function]}
  username="myezasifiso"
>
  <div
    className="SATableDiv"
  >
    <h1>
      Salt Astronomers Liaison Assigning
    </h1>
    <table
      align="center"
      className="SATable"
    >
      <thead>
        <tr>
          <th>
            Proposal Code
          </th>
          <th>
            Summary
          </th>
          <th>
            Proposal Title
          </th>
          <th>
            Proposal Investigator
          </th>
          <th>
            Liaison Astronomer
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          key="liaison-2015-2-MLT-006"
        >
          <td>
            <a
              href="https://www.salt.ac.za/wm/proposal/2015-2-MLT-006"
              target="_blank"
            >
              2015-2-MLT-006
            </a>
          </td>
          <td
            className="width-100"
          >
            <a
              className="file-download"
              href=""
              onClick={[Function]}
            >
              Download
            </a>
          </td>
          <td
            className=" table-height width-400"
          >
            THE SALT GRAVITATIONAL LENSING LEGACY SURVEY
          </td>
          <td
            className="width-100"
          >
            Serjeant Stephen
          </td>
          <td>
            <select
              className="setLiaison"
              defaultValue="brent"
              name="selector"
              onChange={[Function]}
              style={
                Object {
                  "color": "blue",
                }
              }
            >
              <option
                key="brent"
                name="brent"
                value="brent"
              >
                brent
              </option>
              <option
                key="encarni"
                name="encarni"
                value="encarni"
              >
                encarni
              </option>
            </select>
          </td>
        </tr>
        <tr
          key="liaison-2018-1-SCI-038"
        >
          <td>
            <a
              href="https://www.salt.ac.za/wm/proposal/2018-1-SCI-038"
              target="_blank"
            >
              2018-1-SCI-038
            </a>
          </td>
          <td
            className="width-100"
          >
            <a
              className="file-download"
              href=""
              onClick={[Function]}
            >
              Download
            </a>
          </td>
          <td
            className=" table-height width-400"
          >
            Surface Chemistry of Peculiar Hot Subdwarfs
          </td>
          <td
            className="width-100"
          >
            Jeffery Simon
          </td>
          <td>
            <select
              className="setLiaison"
              defaultValue={null}
              name="selector"
              onChange={[Function]}
              style={
                Object {
                  "color": "blue",
                }
              }
            >
              <option>
                none
              </option>
              <option
                key="brent"
                name="brent"
                value="brent"
              >
                brent
              </option>
              <option
                key="encarni"
                name="encarni"
                value="encarni"
              >
                encarni
              </option>
            </select>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</LiaisonTable>
`;

exports[`LiaisonTable Component Render the LiaisonTable Component having unpopulated props with no errors 1`] = `
<LiaisonTable
  canAssign={false}
  initProposals={Array []}
  proposals={Array []}
  requestSummary={[Function]}
  selectArray={Array []}
  semester=""
  setLiaison={[Function]}
  username=""
>
  <div
    className="SATableDiv"
  >
    <h1>
      Salt Astronomers Liaison Assigning
    </h1>
    <table
      align="center"
      className="SATable"
    >
      <thead>
        <tr>
          <th>
            Proposal Code
          </th>
          <th>
            Summary
          </th>
          <th>
            Proposal Title
          </th>
          <th>
            Proposal Investigator
          </th>
          <th>
            Liaison Astronomer
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</LiaisonTable>
`;
